jina/peapods/peas/__init__.py

```diff
@@ -3,22 +3,102 @@
 import os
 import threading
 import time
-from typing import Any, Tuple
+from typing import Any, Tuple, Union, Dict
 
 from .helper import _get_event, ConditionalEvent
 from ... import __stop_msg__, __ready_msg__, __default_host__, __docker_host__
 from ...enums import PeaRoleType, RuntimeBackendType, SocketType, GatewayProtocolType
 from ...excepts import RuntimeFailToStart, RuntimeRunForeverEarlyError
 from ...helper import typename
-from ...hubble.helper import is_valid_huburi, parse_hub_uri
-from ...hubble.hubapi import resolve_local
+from ...hubble.helper import is_valid_huburi
 from ...hubble.hubio import HubIO
 from ...logging.logger import JinaLogger
 from ...parsers.hubble import set_hub_pull_parser
 
 __all__ = ['BasePea']
 
 
+def run(
+    args: 'argparse.Namespace',
+    name: str,
+    runtime_cls,
+    envs: Dict[str, str],
+    timeout_ctrl: int,
+    zed_runtime_ctrl_address: str,
+    is_started: Union['multiprocessing.Event', 'threading.Event'],
+    is_shutdown: Union['multiprocessing.Event', 'threading.Event'],
+    is_ready: Union['multiprocessing.Event', 'threading.Event'],
+    cancel_event: Union['multiprocessing.Event', 'threading.Event'],
+):
+    """Method representing the :class:`BaseRuntime` activity.
+
+    This method is the target for the Pea's `thread` or `process`
+
+    .. note::
+        :meth:`run` is running in subprocess/thread, the exception can not be propagated to the main process.
+        Hence, please do not raise any exception here.
+
+    .. note::
+        Please note that env variables are process-specific. Subprocess inherits envs from
+        the main process. But Subprocess's envs do NOT affect the main process. It does NOT
+        mess up user local system envs.
+
+    .. warning::
+        If you are using ``thread`` as backend, envs setting will likely be overidden by others
+
+    :param args: namespace args from the Pea
+    :param name: name of the Pea to have proper logging
+    :param runtime_cls: the runtime class to instantiate
+    :param envs: a dictionary of environment variables to be set in the new Process
+    :param timeout_ctrl: timeout time for the control port communication
+    :param zed_runtime_ctrl_address: the control address of the `ZEDRuntime` that is supported by the Pea or `ContainerRuntime` or `JinadRuntime`.
+    :param is_started: concurrency event to communicate runtime is properly started. Used for better logging
+    :param is_shutdown: concurrency event to communicate runtime is terminated
+    :param is_ready: concurrency event to communicate runtime is ready to receive messages
+    :param cancel_event: concurrency event to receive cancelling signal from the Pea. Needed by some runtimes
+    """
+    logger = JinaLogger(name, **vars(args))
+
+    def _unset_envs():
```

Comment:
function inside function are not picklable. This will fail during `spawn` context.

Reply:
this does not need to be pickled anymore, can u try  urself?

Reply:
Tested on local. `spawn` works.